onnxruntime/core/framework/session_state.h

```diff
@@ -41,14 +42,23 @@ struct MemoryPatternGroup;
 class TaskThreadPool;
 #endif
 
-// SessionState should be modified by the inference session class only.
-// It is supposed to be passed by const-ref only to all the executors.
+/**
+ * SessionState should be modified by the inference session class only.
+ * It is supposed to be passed by const-ref only to all the executors.
+ * This class owns all the initializers.
+ */
```

Comment:
Does it? We copy the MLValue into initialized_tensors_, and MLValue has a shared_ptr for the data. If there's no deleter, we don't technically own it.

I'd like MLValue to not have a shared_ptr for the data as those ownership semantics seem somewhat loose to me, but that's not how it is currently, and there may be a good reason for it that I'm missing.

Reply:
Whether it has deleter or not, the underlying buffer of the tensors, are owned by this object.  You can't access the initializers after this object has been destroyed.

Reply:
MLValue has a ref count, Tensor class also has a ref count.  So does Allocator. But Yuan suggests we should make lifetime of these objects explicitly simpler. For example, if we clearly knows all the initializers can't live longer than the session, we can remove the ref counting associated with it  **as much as possible**.

Reply:
I agree it should be simpler - my point was (granted somewhat pedantic) more about the comment being misleading. Ownership semantics are important. 

It doesn't really 'own' the initializers, but will ensure they remain valid via the shared_ptr it has to them. If I was new to the code and investigating say a memory issue, I might read the comment and be surprised if initializers weren't freed when SessionState was destructed. If I was a user of onnxruntime and read this I may incorrectly think I'm not able to use the same initializers in different sessions because this would somehow free the ones it was given. I'd need to look at the internals of MLValue and the implementation to realize that wasn't the case.

Reply:
No. It won't ensure that.

---
In reply to: [261699172](https://github.com/Microsoft/onnxruntime/pull/520#discussion_r261699172) [](ancestors = 261699172)